dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/expression/PredictorPreprocessorExpression.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2004-2022, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.expression;
+
+import static java.lang.String.format;
+
+import lombok.Getter;
+
+import org.hisp.dhis.antlr.ParserException;
+
+/**
+ * Parser for predictor expression preprocessing.
+ *
+ * @author Jim Grace
+ */
+@Getter
+public class PredictorPreprocessorExpression
+    extends PreprocessorExpression
```

Comment:
I don't think we should start with using inheritence here. I'd join these to into just `PredictorExpression`. That these can have a preprocessor part is a detail. I'm not sure we want to can share any of that logic with other usages should they arise. We should have a look at that if we get there.

Reply:
OK, done that way. There is now just a single `PredictorExpression` class which holds both preprocessor predictor expressions and non-preprocessor predictor expressions, so it is much less awkward and easier to use.